File: internal/backup/destination.go
package backup

import "github.com/bulletproof-bot/backup/internal/types"

// Destination is an abstract interface for backup destinations
type Destination interface {
	Save(sourcePath string, snapshot *types.Snapshot, message string) error
	GetLastSnapshot() (*types.Snapshot, error)
	GetSnapshot(id string) (*types.Snapshot, error)
	ListSnapshots() ([]*types.SnapshotInfo, error)
	Restore(snapshotID string, targetPath string) error
	Validate() error
	// GetSnapshotPath returns the filesystem path where a snapshot's files are stored
	// Returns empty string if not applicable (e.g., git destination)
	GetSnapshotPath(id string) string
	// DeleteSnapshot deletes a snapshot by ID
	DeleteSnapshot(id string) error
}

File: lib/src/destinations/destination.dart
import '../backup/snapshot.dart';
import '../backup/backup_engine.dart';

/// Abstract interface for backup destinations.
/// 
/// Implementations handle storing and retrieving backups
/// in different formats (git, local folders, cloud sync folders).
abstract class Destination {
  /// Save a backup to this destination
  Future<void> save({
    required String sourcePath,
    required Snapshot snapshot,
    required String message,
  });

  /// Get the most recent snapshot
  Future<Snapshot?> getLastSnapshot();

  /// Get a specific snapshot by ID
  Future<Snapshot?> getSnapshot(String id);

  /// List all available snapshots
  Future<List<SnapshotInfo>> listSnapshots();

  /// Restore files from a snapshot
  Future<void> restore({
    required String snapshotId,
    required String targetPath,
  });

  /// Check if this destination is properly configured
  Future<bool> validate();
}

